internal/expr/typeinfo.go

```diff
@@ -109,9 +112,59 @@ func getTypeInfo(value any) (typeInfo, error) {
 	return typeInfo, nil
 }
 
+// getStructFields requires the caller to check that t is a struct.
+func getStructFields(t reflect.Type) ([]*structField, error) {
+	var fields []*structField
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		if !f.IsExported() {
+			return nil, fmt.Errorf("field %q of struct %s not exported", f.Name, t.Name())
+		}
+		ft := f.Type
+		k := ft.Kind()
+		// Check if the field type is an embedded/nested struct or pointer to
+		// one. If a pointer to the field type implementes Scanner then it is
+		// not a nested struct.
+		if (k == reflect.Struct && !reflect.PointerTo(ft).Implements(scannerInterface)) ||
```

Comment:
Is my understanding correct: Many Go builtin/base types have scanner interface implementations provided by database/sql, but in general, unless explicitly provided, custom structs do not have them implemented? But what happens if a struct type (or pointer to) does provide this interface?

Reply:
This is the case for the `sql.NullString`/`sql.NullInt64`/... types that are in some of the tests.